kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs

```diff
@@ -303,38 +303,48 @@ impl<'tcx> GotocCtx<'tcx> {
+                if adt.is_struct() {
+                    // In this case, we must have a one variant ADT. There are two cases.
+                    let variant = adt.non_enum_variant();
+                    let overall_type = self.codegen_ty(ty);
+                    // If there is no field, then it's just a ZST.
                     if variant.fields.is_empty() {
-                        if adt.is_struct() {
-                            let overall_t = self.codegen_ty(ty);
-                            Expr::struct_expr_from_values(overall_t, vec![], &self.symbol_table)
-                        } else {
-                            unimplemented!()
-                        }
+                        Expr::struct_expr_from_values(overall_type, vec![], &self.symbol_table)
+                    // Otherwise, there is at least one field associated with the scalar data.
+                    // Any additional fields correspond to ZSTs.
                     } else {
-                        // otherwise, there is just one field, which is stored as the scalar data
-                        let field = &variant.fields[0usize.into()];
-                        let fty = field.ty(self.tcx, subst);
-
-                        let overall_t = self.codegen_ty(ty);
-                        if adt.is_struct() {
-                            self.codegen_single_variant_single_field(s, span, overall_t, fty)
-                        } else {
-                            unimplemented!()
-                        }
+                        let field_types: Vec<Ty<'_>> =
+                            variant.fields.iter().map(|f| f.ty(self.tcx, subst)).collect();
+                        // Check that there is a single non-ZST field.
+                        let non_zst_types: Vec<_> =
+                            field_types.iter().filter(|t| !self.is_zst(**t)).collect();
+                        assert!(
+                            non_zst_types.len() == 1,
+                            "error: expected exactly one field whose type is not a ZST"
+                        );
+                        let field_values: Vec<Expr> = field_types
+                            .iter()
+                            .map(|t| {
+                                if self.is_zst(*t) {
+                                    Expr::init_unit(self.codegen_ty(*t), &self.symbol_table)
+                                } else {
+                                    self.codegen_scalar(s, *t, span)
+                                }
+                            })
+                            .collect();
+                        Expr::struct_expr_from_values(
+                            overall_type,
+                            field_values,
+                            &self.symbol_table,
+                        )
                     }
-                } else {
-                    // if it's an enum
+                } else if adt.is_enum() {
```

Comment:
can we have a case where all enums have zero sized fields and the scalar value returned by the constant evaluation represents the discriminant of the enum ?

Reply:
Sorry, I don't understand the question.

This PR does not change how we generate code for ADTs represented with `enum`s, only ADTs which are `struct`s because that's when the crash happened. The change you highlighted comes from the branch clean-up.